cloudferrylib/os/actions/filter_similar_vms_from_dst.py

```diff
@@ -0,0 +1,89 @@
+        self.tenant_id_to_new_id = {}
+        for tenant in kwargs['identity_info']['tenants']:
+            self.tenant_id_to_new_id[tenant['tenant']['id']] = \
+                tenant['meta']['new_id']
+        self.find_simiral_instances()
+        for src_instance_id, dst_ids in self.similar_isntances.items():
+            LOG.warning("Instance %s already in DST cloud as instance %s. "
+                        "It will be excluded from migration." %
+                        (src_instance_id, dst_ids))
+            self.src_instances.pop(src_instance_id)
+        for src_instance_id, dst_ids in self.conflict_instances.items():
+            LOG.warning("Instance %s can not be migrated to DST because "
+                        "instance %s already use the same IP. "
+                        "It will be excluded from migration." %
+                        (src_instance_id, dst_ids))
+            self.src_instances.pop(src_instance_id)
+
+    def find_simiral_instances(self):
+        # titii = tenant & ip to instance id
+        titii_src = self.make_tenant_ip_to_instance_id_dict(self.src_instances)
+        search_opts = {'search_opts_tenant': titii_src.keys()}
+        compute_resource = self.dst_cloud.resources[utils.COMPUTE_RESOURCE]
+        dst_instances = compute_resource.read_info(**search_opts)['instances']
+        self.dst_instances = dst_instances
+        titii_dst = self.make_tenant_ip_to_instance_id_dict(dst_instances)
+        self.similar_isntances = collections.defaultdict(set)
+        self.conflict_instances = collections.defaultdict(set)
+        for tenant_id, ip_to_id in titii_src.items():
+            tenant_new_id = self.tenant_id_to_new_id[tenant_id]
+            dst_ip_to_id = titii_dst[tenant_new_id]
+            for ip, instance_id in ip_to_id.items():
+                if ip in dst_ip_to_id:
+                    self.instance_comparison(instance_id, dst_ip_to_id[ip])
+
+    @staticmethod
+    def make_tenant_ip_to_instance_id_dict(instances):
+        tenant_ip_to_instance_id = collections.defaultdict(dict)
+        for instance in instances.values():
+            info = instance['instance']
+            ip_to_id = tenant_ip_to_instance_id[info['tenant_id']]
+            instance_id = info['id']
+            for interface in info['interfaces']:
+                ip_to_id[interface['ip']] = instance_id
+        return tenant_ip_to_instance_id
+
+    def instance_comparison(self, src_instance_id, dst_instance_id):
+        src_instance = self.src_instances[src_instance_id]['instance']
+        dst_instance = self.dst_instances[dst_instance_id]['instance']
+        equal = True
+        for field in COMPARISON_FIELDS:
+            if src_instance[field] != dst_instance[field]:
+                equal = False
+                break
+        if equal:
+            self.similar_isntances[src_instance_id].add(dst_instance_id)
+        else:
+            self.conflict_instances[src_instance_id].add(dst_instance_id)
```

Comment:
How about introducing `VM` class with `__eq__` method defined instead all of this logic?..